xonsh/prompt/env.py

```diff
@@ -1,39 +1,55 @@
 """Prompt formatter for virtualenv and others"""
 
-import os
+import re
+from pathlib import Path
 
 from xonsh.built_ins import XSH
 
 
-def find_env_name():
-    """Finds the current environment name from $VIRTUAL_ENV or
-    $CONDA_DEFAULT_ENV if that is set.
+def env_name():
+    """Build env_name based on different sources. Respect order of precedence.
+
+    Name from VIRTUAL_ENV_PROMPT will be used as-is.
+    Names from other sources are surrounded with ``{env_prefix}`` and
+    ``{env_postfix}`` fields.
     """
-    env_path = XSH.env.get("VIRTUAL_ENV", "")
-    if env_path:
-        env_name = os.path.basename(env_path)
-    else:
-        env_name = XSH.env.get("CONDA_DEFAULT_ENV", "")
-    return env_name
+    if XSH.env.get("VIRTUAL_ENV_DISABLE_PROMPT"):
+        return
+    virtual_env_prompt = XSH.env.get("VIRTUAL_ENV_PROMPT")
+    if virtual_env_prompt:
+        return virtual_env_prompt
+    virtual_env = XSH.env.get("VIRTUAL_ENV")
+    if virtual_env:
+        venv_path = Path(virtual_env)
+        pyvenv_cfg_prompt = prompt_from_pyvenv_cfg(venv_path)
+        if pyvenv_cfg_prompt:
+            return surround_env_name(pyvenv_cfg_prompt)
+        return surround_env_name(venv_path.name)
+    from_conda = XSH.env.get("CONDA_DEFAULT_ENV")
+    if from_conda:
+        return surround_env_name(from_conda)
+    return
 
 
-def env_name():
-    """Returns the current env_name if it non-empty, surrounded by the
-    ``{env_prefix}`` and ``{env_postfix}`` fields.
+def prompt_from_pyvenv_cfg(venv_path):
+    """Grab the prompt from the venv configuration, if it exists.
+
+    Tries to be resilient to subtle changes in whitespace and quoting in the
+    configuration file format as it adheres to no clear standard.
     """
-    env_name = find_env_name()
-    if XSH.env.get("VIRTUAL_ENV_DISABLE_PROMPT") or not env_name:
-        # env name prompt printing disabled, or no environment; just return
-        return
+    assert isinstance(venv_path, Path), venv_path
```

Comment:
can the assert statement be skipped?

Reply:
You mean remove the assert? I usually like to be a bit more paranoid, especially in highly dynamic code without type hinting. It also helps IDEs to figure out, what is going on. But if this against the grain of this project, I can remove it. If it is again about speed, this problem would als be solved, if the prompt is stored in VIRTUAL_ENV_PROMPT after first access.

Reply:
well that can be an if statement. We use mypy, and try to add type hint wherever possible.

Reply:
The assert has lost its usefulness in the new version anyway, so it's gone. But good to know that you use type hints. In the parts of the code I was until now.